h2o-core/src/main/java/water/parser/CsvParser.java

```diff
@@ -149,21 +157,24 @@
             str.addBuff(bits);
           }
           if( !isNa &&
-              _setup.isNA(colIdx, str.toBufferedString())) {
+              _setup.isNA(columnCounter, str.toBufferedString())) {
             isNa = true;
           }
-          if (!isNa) {
-            dout.addStrCol(colIdx, str.toBufferedString());
+
+          if (!isNa && (colIdx <= colIndexNum) && _keepColumns[colIdx]) {
```

Comment:
I am not sure if the condition `(colIdx <= colIndexNum)` isn't in vain ? colIdx will always be at most the number of columns minus one in the original CSV file, as `int colIndexNum = _keepColumns.length-1;`.

This condition occurs on multiple lines.

Reply:
Pavel:

I agreed with you at first.  However, after looking at this and checking out the runs in detail, this condition is essential.  Basically, during the run, the colIdx will be increased beyond colIndexNum if I do not put the condition in there.  It is intentionally designed by Tomas.

I used to have colIdx < colIndexNum but this creates the error of doubling the number of rows of certain files.  This test will fail:

Java 10 JUnit / water.parser.ParserTest.testParserRespectsSpecifiedColNum.

After checking the failure, I found the problem is that I need to let colIdx increment beyond its legal limit: colIdx == colIndexNum.  The reason for colIdx to increase beyond the actual number of columns is due to the finite state machine that Tomas has designed.  An extra row will be added if the condition colIdx < colIndexNum is satisfied.  However, this extra row will not be added if we have colIdx >= colIndexNum.

Reply:
I see. Did not realize there is an extra row at the end. You're right Wendy.

Reply:
Pavel:  As an experiment, you can change the condition from colIdx <= colIndexNum to colIdx < colIndexNum and see the doubling of the number of rows in that test.